tests/test_source.py

```diff
@@ -0,0 +1,84 @@
+import logging
+
+import pytest
+from unittest.mock import MagicMock
+
+from src.arcam.fmj.client import Client
+from src.arcam.fmj.state import State
+from src.arcam.fmj import (
+  SourceCodes,
+  ApiModel,
+  AnswerCodes,
+  CommandCodes,
+  ResponsePacket,
+  SOURCE_CODES,
+  SOURCE_WRITE_SUPPORTED,
+  RC5CODE_SOURCE,
+)
+
+NO_SOURCE = {ApiModel.APIPA_SERIES}
+
+TEST_PARAMS = [
+    (1, ApiModel.API450_SERIES),
+    (1, ApiModel.API860_SERIES),
+    (1, ApiModel.APIHDA_SERIES),
+    (1, ApiModel.APISA_SERIES),
+    (1, ApiModel.APIPA_SERIES),
```

Comment:
for testing the PA series i had checked if an exception was raised because that model does not support source